Mostly static binaries

We already support building static libraries. In fact we already know how to static binaries as well: when profiling is enabled we use static linking because profiling and dynamic linking don't mix. This ticket is about exposing to the user the functionality, so that users can *force* static linking.

This ticket is about static linking in the sense of GHC: in Bazel terms, this corresponds to "mostly static" linking. That is, Haskell code is statically linked, but system libraries (like libgmp, glibc, libpthread etc) are still dynamically linked.

cc @lunaris

I will try and get a patch going for this, because simply uncommenting the `elif not with_profiling` part of `haskell_link_binary` that adds `-pie`-etc. arguments seems to work just fine:

```
➜  test git:(bazel) ✗ ldd bazel-bin/engine/bin/static-binary/static-binary                                    
        linux-vdso.so.1 =>  (0x00007ffc9999f000)
        libstdc++.so.6 => /nix/store/zk5zj2307zxaq7dx585yia3dn5k4qlsl-gcc-7.3.0-lib/lib/libstdc++.so.6 (0x00007f84f5e39000)
        libm.so.6 => /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/libm.so.6 (0x00007f84f5aa4000)
        libssh2.so.1 => /nix/store/n6i3f9q2r1ikbdhy6j51c30fslrhqvww-libssh2-1.8.0/lib/libssh2.so.1 (0x00007f84f5876000)
        libpcre.so.1 => /nix/store/3jg1sj3va3rmm4fhw5xkybq9q3v8wnyi-pcre-8.42/lib/libpcre.so.1 (0x00007f84f5604000)
        liblua.so.5.3 => /nix/store/m7bphi4y05byj5kqiafgqx19y6biv4wb-lua-5.3.5/lib/liblua.so.5.3 (0x00007f84f53cb000)
        libz.so.1 => /nix/store/bv6znzsv2qkbcwwa251dx7n5dshz3nr3-zlib-1.2.11/lib/libz.so.1 (0x00007f84f51af000)
        libpthread.so.0 => /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/libpthread.so.0 (0x00007f84f4f90000)
        libpq.so.5 => /nix/store/1206craj2bvn18slflp125mkh8ncp5dz-postgresql-9.6.10-lib/lib/libpq.so.5 (0x00007f84f4d61000)
        librt.so.1 => /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/librt.so.1 (0x00007f84f4b59000)
        libutil.so.1 => /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/libutil.so.1 (0x00007f84f4956000)
        libdl.so.2 => /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/libdl.so.2 (0x00007f84f4752000)
        libgmp.so.10 => /nix/store/dsjjfh28bqg78773r7zvq33giwsbyiqb-gmp-6.1.2/lib/libgmp.so.10 (0x00007f84f44be000)
        libgcc_s.so.1 => /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/libgcc_s.so.1 (0x00007f84f42a8000)
        libc.so.6 => /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/libc.so.6 (0x00007f84f3ef4000)
        /nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/ld-linux-x86-64.so.2 => /lib64/ld-linux-x86-64.so.2 (0x00007f84f61c0000)
        libssl.so.1.0.0 => /nix/store/gsgdj34vv5hh3iyyfm7yh69j6dq0br9q-openssl-1.0.2p/lib/libssl.so.1.0.0 (0x00007f84f3c80000)
        libcrypto.so.1.0.0 => /nix/store/gsgdj34vv5hh3iyyfm7yh69j6dq0br9q-openssl-1.0.2p/lib/libcrypto.so.1.0.0 (0x00007f84f3819000)

```

Fully-static however seems like it will be much more problematic -- adding the traditional `-optl-static -optl-pthread` to `compiler_flags` for a given `haskell_binary` produces lots of errors like:

```
...
/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib/gcc/x86_64-unknown-linux-gnu/7.3.0/../../../../lib64/libstdc++.a(guard.o):function __cxa_guard_abort: error: undefined reference to 'syscall'
/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib/gcc/x86_64-unknown-linux-gnu/7.3.0/../../../../lib64/libstdc++.a(guard.o):function __cxa_guard_release: error: undefined reference to 'syscall'
/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib/gcc/x86_64-unknown-linux-gnu/7.3.0/../../../../lib64/libstdc++.a(vterminate.o):function __gnu_cxx::__verbose_terminate_handler(): error: undefined reference to 'fputs'
/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib/gcc/x86_64-unknown-linux-gnu/7.3.0/../../../../lib64/libstdc++.a(vterminate.o):function __gnu_cxx::__verbose_terminate_handler(): error: undefined reference to 'fputs'
/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib/gcc/x86_64-unknown-linux-gnu/7.3.0/../../../../lib64/libstdc++.a(cp-demangle.o):function d_append_num: error: undefined reference to 'sprintf'
/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib/gcc/x86_64-unknown-linux-gnu/7.3.0/../../../../lib64/libstdc++.a(cp-demangle.o):function d_print_comp_inner: error: undefined reference to 'sprintf'
/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib/gcc/x86_64-unknown-linux-gnu/7.3.0/../../../../lib64/libstdc++.a(cp-demangle.o):function d_print_comp_inner: error: undefined reference to 'sprintf'
/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib/gcc/x86_64-unknown-linux-gnu/7.3.0/../../../../lib64/libstdc++.a(cp-demangle.o):function d_print_comp_inner: error: undefined reference to 'sprintf'
/nix/store/vv4r320p5yd1k01kld62q1lppjxcswhb-gcc-7.3.0/lib/gcc/x86_64-unknown-linux-gnu/7.3.0/libgcc_eh.a(unwind-dw2-fde-dip.o):function _Unwind_Find_FDE: error: undefined reference to 'dl_iterate_phdr'
collect2: error: ld returned 1 exit status
`cc' failed in phase `Linker'. (Exit code: 1)
Target //engine/bin/static-binary:static-binary failed to build
INFO: Elapsed time: 30.358s, Critical Path: 29.15s
INFO: 1 process: 1 linux-sandbox.

```

--- a/haskell/haskell.bzl
+++ b/haskell/haskell.bzl
@@ -107,6 +107,10 @@ def _mk_binary_rule(**kwargs):
         executable = True,
         attrs = dict(
             _haskell_common_attrs,
+            linkstatic = attr.bool(
+                default = False,
+                doc = "If enabled, this option tells GHC to link statically whenever possible. While this captures all Haskell code built, some system libraries may still be linked dynamically, as are libraries for which there is no static library. So the resulting executable will still be dynamically linked, hence only mostly static.",
+            ),
             generate_so = attr.bool(
                 default = False,
                 doc = "Whether to generate also a .so version of executable.",

--- a/haskell/private/actions/link.bzl
+++ b/haskell/private/actions/link.bzl
@@ -122,14 +122,6 @@ def link_binary(
     if not linkstatic:
         if hs.toolchain.is_darwin:
             args.add(["-optl-Wl,-headerpad_max_install_names"])
-
-            # Nixpkgs commit 3513034208a introduces -liconv in NIX_LDFLAGS on
-            # Darwin. We don't currently handle NIX_LDFLAGS in any special
-            # way, so a hack is to simply do what NIX_LDFLAGS is telling us we
-            # should do always when using a toolchain from Nixpkgs.
-            # TODO remove this gross hack.
-            # TODO: enable dynamic linking of Haskell dependencies for macOS.
-            args.add("-liconv")
         elif not with_profiling:
             args.add(["-pie", "-dynamic"])
 
@@ -168,6 +160,14 @@ def link_binary(
             "-optc-Wno-unused-command-line-argument",
             "-optl-Wno-unused-command-line-argument",
         ])
+
+        # Nixpkgs commit 3513034208a introduces -liconv in NIX_LDFLAGS on
+        # Darwin. We don't currently handle NIX_LDFLAGS in any special
+        # way, so a hack is to simply do what NIX_LDFLAGS is telling us we
+        # should do always when using a toolchain from Nixpkgs.
+        # TODO remove this gross hack.
+        # TODO: enable dynamic linking of Haskell dependencies for macOS.
+        args.add("-liconv")
     else:
         for rpath in set.to_list(_infer_rpaths(executable, solibs)):
             args.add(["-optl-Wl,-rpath," + rpath])

--- a/haskell/private/actions/link.bzl
+++ b/haskell/private/actions/link.bzl
@@ -83,6 +83,7 @@ def link_binary(
         extra_srcs,
         compiler_flags,
         objects_dir,
+        linkstatic,
         with_profiling,
         dummy_static_lib,
         version):
@@ -111,18 +112,26 @@ def link_binary(
     args.add(hs.toolchain.compiler_flags)
     args.add(compiler_flags)
 
-    if hs.toolchain.is_darwin:
-        args.add(["-optl-Wl,-headerpad_max_install_names"])
+    # By default, GHC will produce mostly-static binaries, i.e. in which all
+    # Haskell code is statically linked and foreign libraries and system
+    # dependencies are dynamically linked. If linkstatic is false, i.e. the user
+    # has requested fully dynamic linking, we must therefore add flags to make
+    # sure that GHC dynamically links Haskell code too. The one exception to
+    # this is when we are compiling for profiling, which currently does not play
+    # nicely with dynamic linking.
+    if not linkstatic:
+        if hs.toolchain.is_darwin:
+            args.add(["-optl-Wl,-headerpad_max_install_names"])
 
-        # Nixpkgs commit 3513034208a introduces -liconv in NIX_LDFLAGS on
-        # Darwin. We don't currently handle NIX_LDFLAGS in any special
-        # way, so a hack is to simply do what NIX_LDFLAGS is telling us we
-        # should do always when using a toolchain from Nixpkgs.
-        # TODO remove this gross hack.
-        # TODO: enable dynamic linking of Haskell dependencies for macOS.
-        args.add("-liconv")
-    elif not with_profiling:
-        args.add(["-pie", "-dynamic"])
+            # Nixpkgs commit 3513034208a introduces -liconv in NIX_LDFLAGS on
+            # Darwin. We don't currently handle NIX_LDFLAGS in any special
+            # way, so a hack is to simply do what NIX_LDFLAGS is telling us we
+            # should do always when using a toolchain from Nixpkgs.
+            # TODO remove this gross hack.
+            # TODO: enable dynamic linking of Haskell dependencies for macOS.
+            args.add("-liconv")
+        elif not with_profiling:
+            args.add(["-pie", "-dynamic"])
 
     args.add(["-o", compile_output.path, dummy_static_lib.path])
 

--- a/haskell/private/haskell_impl.bzl
+++ b/haskell/private/haskell_impl.bzl
@@ -111,6 +111,7 @@ use the 'haskell_import' rule instead.
         ctx.files.extra_srcs,
         ctx.attr.compiler_flags,
         c_p.objects_dir if with_profiling else c.objects_dir,
+        ctx.attr.linkstatic,
         with_profiling,
         ctx.file._dummy_static_lib,
         version = ctx.attr.version,

--- a/tests/binary-mostly-static/BUILD
+++ b/tests/binary-mostly-static/BUILD
@@ -1,0 +1,14 @@
+package(default_testonly = 1)
+
+load(
+    "@io_tweag_rules_haskell//haskell:haskell.bzl",
+    "haskell_binary",
+)
+
+haskell_binary(
+    name = "binary-mostly-static",
+    srcs = ["Main.hs"],
+    visibility = ["//visibility:public"],
+    deps = ["//tests:base"],
+    linkstatic = True,
+)

--- a/tests/binary-mostly-static/Main.hs
+++ b/tests/binary-mostly-static/Main.hs
@@ -1,0 +1,3 @@
+module Main where
+
+main = putStrLn "hello world"

